docs/ocis/adr/0029-grpc-in-kubernetes.md

```diff
@@ -0,0 +1,102 @@
+---
+title: "29. gRPC in Kubernetes"
+date: 2024-06-27T14:05:00+01:00
+weight: 29
+geekdocRepo: https://github.com/owncloud/ocis
+geekdocEditPath: edit/master/docs/ocis/adr
+geekdocFilePath: 0029-grpc-in-kubernetes.md
+---
+
+* Status: draft
+* Deciders: [@butonic](https://github.com/butonic)
+* Date: 2024-06-27
+
+## Context and Problem Statement
+
+[Scaling oCIS in kubernetes causes requests to fail
+#8589](https://github.com/owncloud/ocis/issues/8589), sometimes until affected services have manually been restarted. This can be observed by two symptoms:
+1. when a new pod is added it does not receive traffic
+2. when an pod is shut down existing clients still try to send requests to it
+
+To leverage the kubernetes pod state, we first used the go micro kubernetes registry implementation. When a pod fails the health or readyness probes, kubernetes will no longer
+- send traffic to the pod via the kube-proxy, which handles the ClusterIP for a service, 
+- list the pod in DNS responses when the ClusterIP is disabled by setting it to `none`
+When using the ClusterIP HTTP/1.1 requests will be routed to a working pod. 
+
+This nice setup starts to fail with long lived connections. The kube-proxy is connection based, causing requests with Keep-Alive to stick to the same pod for more than one request. Worse, HTTP/2 and in turn gRPC are multiplexing the connection. They will not pick up any changes to pods, explaining the symptomps:
```

Comment:
are we using this at all? Isn't the Go micro registry returning pod IPs? 
And from what I know, the nats-js-kv service registry doesn't have any insights into healthiness or readiness of services it tries to contact.

Reply:
sorry, have seen it in https://github.com/owncloud/ocis/pull/9488/files#diff-0701d2d4266b791c472766e356162c2ea1810b1037d5036c36b888ff9f947ab8R36